### ISSUE TITLE
Performance regression in unsafe_getindex

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#14594]
I looked at https://github.com/jrevels/BaseBenchmarkReports/blob/master/c70ab26/c70ab26_vs_adffe19.md which is a test benchmark run that @jrevels ran that compares 0.4.2 with master. There are multiple regressions and I believe that most of them are from changes to `unsafe_getindex`, and more specifically, `_checksize`

For example, the benchmark:
```jl
function sumvector(A, n)
    s = zero(eltype(A)) + zero(eltype(A))
    nrows = size(A, 1)
    ncols = size(A, 2)
    r = rand(1:nrows, 5)
    for k = 1:n
        for i = 1:ncols
            val = Base.unsafe_getindex(A, r, i)
            s += first(val)
        end
    end
    s
end
```

Benchmark with `A = rand(300, 500)`.

Master:

```jl
julia> @time sumvector(A, 1000);
  0.126164 seconds (2.00 M allocations: 83.924 MB, 6.93% gc time)
```

0.4.2:
```jl
julia> @time sumvector(A, 1000)
  0.034920 seconds (1.00 M allocations: 61.035 MB, 7.24% gc time)
```

Profiling shows that the extra time is spend in `_checksize` at https://github.com/JuliaLang/julia/blob/128e546a4318f1908398e02828781315cae95a50/base/multidimensional.jl#L259

CC: @mbauman since it was last touched in the scalar dimension drop change.

**Comment by mbauman** [comment]
That benchmark tracker is already proving to be invaluable.  Thanks for making this issue, @KristofferC.  I'll try to look into this when I have time.

**Comment by mbauman** [comment]
Fixed by #14609 and superseded by #14651 for the remaining regressions.
